### PR TITLE
Enable DNN supercluster in Phase2 HLT

### PR DESCRIPTION
**Pull request by RSalvatico**
#### PR description:

Enable the usage of the DNN superclustering developed for the endcaps (PR #45333) in the Phase2 HLT reconstruction. In particular, this PR allows us to use:

- Superclustering DNN (TracksterLinkingBySuperClusteringDNN)
- Superclustering with Mustache inside TICL (TracksterLinkingbySuperClusteringMustache)

using process modifiers.

#### PR validation:

Successfully run several upgrade workflows such as 
29691.203 SingleElectronPt15Eta1p7_2p7_2026D110 + ticl_v5
29691.204 SingleElectronPt15Eta1p7_2p7_2026D110 + ticl_v5_mustache

### CODE REVIEW

**Comment by felicepantaleo** [comment]
test parameters:

- workflow_opts= -w upgrade
- workflows = 29888.203,29688.203

**Comment by felicepantaleo** [comment]
@cmsbuild please test

**Comment by felicepantaleo** [comment]
@DebabrataBhowmik @waredjeb @rovere @VourMa fyi

**Review by mmusich — COMMENTED** [review]
cosmetics

File: HLTrigger/Configuration/python/HLT_75e33/sequences/HLTHgcalTiclPFClusteringForEgammaUnseededSequence_cfi.py [hunk 1]
```diff
@@ -18,13 +18,137 @@
...
+                        + hltHGCalRecHit+hltParticleFlowRecHitHGC
+                        + hltHgcalSoARecHitsProducer
+                        + hltHgcalSoARecHitsLayerClustersProducer
+                        + hltHgcalSoALayerClustersProducer
+                        + hltHgCalLayerClustersFromSoAProducer
+                        + hltHgcalLayerClustersHSci
+                        + hltHgcalLayerClustersHSi
+                        + hltHgcalMergeLayerClusters
+                        ) 
+)
```
Comment (VourMa) [inline HLTrigger/Configuration/python/HLT_75e33/sequences/HLTHgcalTiclPFClusteringForEgammaUnseededSequence_cfi.py:112]: How does this differ from 
https://github.com/cms-sw/cmssw/blob/a85bda57ecb39ceb07427c917c4d4dc63594c9cb/HLTrigger/Configuration/python/HLT_75e33/sequences/HLTHgcalTiclPFClusteringForEgammaUnseededSequence_cfi.py#L30-L45?

Don't the aforementioned lines cover the heterogeneous workflow more in general? Maybe I am missing how the `ticl_superclustering_dnn` modifier changes anything here.

File: HLTrigger/Configuration/python/HLT_75e33/sequences/HLTHgcalTiclPFClusteringForEgammaUnseededSequence_cfi.py [hunk 2]
```diff
@@ -18,13 +18,137 @@
...
```
Comment (VourMa) [inline HLTrigger/Configuration/python/HLT_75e33/sequences/HLTHgcalTiclPFClusteringForEgammaUnseededSequence_cfi.py:141]: Same comment as above.

File: HLTrigger/Configuration/python/HLT_75e33/sequences/HLTHgcalTiclPFClusteringForEgammaUnseededSequence_cfi.py [hunk 3]
```diff
@@ -18,13 +18,137 @@
...
+(alpaka & ticl_superclustering_dnn).toModify(hltHgcalMergeLayerClusters,
+                                             layerClustersEE = cms.InputTag("hltHgCalLayerClustersFromSoAProducer"),
+                                             time_layerclustersEE = cms.InputTag("hltHgCalLayerClustersFromSoAProducer", "timeLayerCluster"))
```
Comment (VourMa) [inline HLTrigger/Configuration/python/HLT_75e33/sequences/HLTHgcalTiclPFClusteringForEgammaUnseededSequence_cfi.py:115]: Similar comment as above when compared to lines
https://github.com/cms-sw/cmssw/blob/a85bda57ecb39ceb07427c917c4d4dc63594c9cb/HLTrigger/Configuration/python/HLT_75e33/sequences/HLTHgcalTiclPFClusteringForEgammaUnseededSequence_cfi.py#L46-L48
Similarly applied to 
https://github.com/cms-sw/cmssw/blob/a85bda57ecb39ceb07427c917c4d4dc63594c9cb/HLTrigger/Configuration/python/HLT_75e33/sequences/HLTHgcalTiclPFClusteringForEgammaUnseededSequence_cfi.py#L142-L144

In general the "# Alpaka & DNN" part seems to be duplicating modifications from above, potentially making the maintenance of the code harder. Maybe try to simplify - unless there is some reason for the duplication which I failed to see (potentially having to do with the order the modifiers are applied)?

File: HLTrigger/Configuration/python/HLT_75e33/sequences/HLTHgcalTiclPFClusteringForEgammaUnseededSequence_cfi.py (hunk 1 above)
Comment (RSalvatico) [reply HLTrigger/Configuration/python/HLT_75e33/sequences/HLTHgcalTiclPFClusteringForEgammaUnseededSequence_cfi.py:112]: No, I think you're right. In fact all the `(alpaka & ticl_superclustering)` modifiers are pointless, as `alpaka` and `supercluster` modify independent sequences. Thanks

File: HLTrigger/Configuration/python/HLT_75e33/sequences/HLTHgcalTiclPFClusteringForEgammaUnseededSequence_cfi.py (hunk 3 above)
Comment (RSalvatico) [reply HLTrigger/Configuration/python/HLT_75e33/sequences/HLTHgcalTiclPFClusteringForEgammaUnseededSequence_cfi.py:115]: Yes indeed, I'll clean up.

**Comment by RSalvatico** [comment]
type egamma

**Comment by mmusich** [comment]
test parameters:
   * workflow_opts= -w upgrade
   * workflows = 29691.203, 29691.204

**Comment by mmusich** [comment]
@cmsbuild, please abort

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
> HLT P2 Timing: [chart](https://cmssdt.cern.ch/circles/web/piechart.php?data_name=hlt-p2-timing&resource=time_thread&filter=CMSSW_14_2_X_2024-09-17-2300&dataset=CMSSW_14_2_X_2024-09-17-2300-cmssw46010/el8_amd64_gcc12/21/Phase2Timing_resources)

the phase-2 timing results show an ever so slight increase of timing from **5630.0 ms/ev** to **5641.6 ms/ev** (**+0.2%**) which is not particularly alarming.
On the other hand there are changes in the trigger results for workflows:
   * [29691.203](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-17-2300+eb053e/64587/triggerResults/29691.203_SingleElectronPt15Eta1p7_2p7+2026D110_ticl_v5/HLT.log), where there is a **net gain** of events triggered by `MC_Ele5_Open_*`
   * [29691.204](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-17-2300+eb053e/64587/triggerResults/29691.204_SingleElectronPt15Eta1p7_2p7+2026D110_ticl_v5_mustache/HLT.log) where there is a **net loss** of events triggered by `MC_Ele5_Open_*`

> DQMHistoTests: [Total failures: 5979](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-17-2300+eb053e/64587/validateJR.html)
   
Concerning the DQM comparisons:
   * for [29691.203](https://tinyurl.com/24ac6n2r)  (`ticlv5`) I see *an increase*  of reconstructed gsf tracks at HLT (see https://tinyurl.com/23wcjxyh)
   * for [29691.204](https://tinyurl.com/2dbtfvmr)  (`ticl_v5_mustache`) it seems that there are no gsf tracks reconstructed at HLT at all (see https://tinyurl.com/2xj74ssu)
   
Perhaps I am misunderstanding something, but somethings looks wrong in the   `ticl_v5_mustache` option.
@RSalvatico can you have a look and comment?

**Comment by RSalvatico** [comment]
Hi @mmusich, there are some elements of this report that I don't understand.

1. [29691.203](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-17-2300+eb053e/64587/triggerResults/29691.203_SingleElectronPt15Eta1p7_2p7+2026D110_ticl_v5/HLT.log) says there were no events accepted by `MC_Ele5_Open_L1Seeded` and with `ticl_v5 (dnn)` we gained 8. [29691.204](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-17-2300+eb053e/64587/triggerResults/29691.204_SingleElectronPt15Eta1p7_2p7+2026D110_ticl_v5_mustache/HLT.log) says there were **zero** events triggered by `MC_Ele5_Open_L1Seeded` and with `ticl_v5_mustache` we lost them all. Aren't these reports in contrast? A similar consideration applies to `MC_Ele5_Open_Unseeded`.
2. If with `ticl_v5_mustache` we lose all the GSF tracks, how come the only big trigger differences come from these two open paths? Could be the small number of test events...

Anyway, I'll try to do more tests today. Thanks

**Comment by RSalvatico** [comment]
> > [29691.204](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-17-2300+eb053e/64587/triggerResults/29691.204_SingleElectronPt15Eta1p7_2p7+2026D110_ticl_v5_mustache/HLT.log) says there were zero events triggered by MC_Ele5_Open_L1Seeded and with ticl_v5_mustache we lost them all. Aren't these reports in contrast?
> 
> I am not sure to follow. It says that there were 9 events and you lost all of them.
> 

Sorry, I meant nine, not zero. I edited the previous comment, now my point should be clearer.

**Comment by RSalvatico** [comment]
The reason why we were seeing zero GSF tracks at HLT when using `ticl_v5_mustache` is that the PID that selects tracksters to form superclusters was disabled in a way that no tracksters would be selected - hence no SCs and no related GSF tracks. I have now enabled it in `hltTiclTrackstersCLUE3DHigh(L1Seeded)` .

Concerning the DNN supercluster (standalone `ticl_v5` customization, .203 workflows), the reason why we see an increased number of GSF tracks is that, at the moment, all the tracksters above 4 GeV that didn't contribute to form a SC are promoted to SCs themselves (see [here](https://github.com/cms-sw/cmssw/blob/22badb1b556fda665443627573e8c44ae178f9fa/RecoHGCal/TICL/plugins/TracksterLinkingbySuperClusteringDNN.cc#L320-L327)). When increasing the [seed pt threshold](https://github.com/cms-sw/cmssw/blob/22badb1b556fda665443627573e8c44ae178f9fa/RecoHGCal/TICL/plugins/TracksterLinkingbySuperClusteringDNN.cc#L362), the number of GSF tracks goes down. This is true for the offline reconstruction as well, although it's probably more relevant to HLT due to timing concerns. But at the moment there doesn't seem to be a timing problem.

On a technical note: I do not understand how to resolve the highlighted conflicts. The modification I want to make is exactly what I have pushed and changes the content of the master, so I am not sure where the conflict lies.

**Comment by mmusich** [comment]
> On a technical note: I do not understand how to resolve the highlighted conflicts. The modification I want to make is exactly what I have pushed and changes the content of the master, so I am not sure where the conflict lies.

@RSalvatico I think the conflict developed after https://github.com/cms-sw/cmssw/pull/45821/ was merged.
I don't know if it helps, but trying to merge locally your branch I get this:

```diff
diff --cc HLTrigger/Configuration/python/HLT_75e33/modules/hltTiclTrackstersCLUE3DHighL1Seeded_cfi.py
index 4a5bd170978,e22e5e9df8c..00000000000
--- a/HLTrigger/Configuration/python/HLT_75e33/modules/hltTiclTrackstersCLUE3DHighL1Seeded_cfi.py
+++ b/HLTrigger/Configuration/python/HLT_75e33/modules/hltTiclTrackstersCLUE3DHighL1Seeded_cfi.py
@@@ -141,5 -141,5 +141,9 @@@ hltTiclTrackstersCLUE3DHighL1Seeded = c
  )
  
  from Configuration.ProcessModifiers.ticl_v5_cff import ticl_v5
++<<<<<<< HEAD
 +ticl_v5.toModify(hltTiclTrackstersCLUE3DHighL1Seeded.pluginPatternRecognitionByCLUE3D, computeLocalTime = cms.bool(True), doPidCut = cms.bool(False))
++=======
+ ticl_v5.toModify(hltTiclTrackstersCLUE3DHighL1Seeded.pluginPatternRecognitionByCLUE3D, computeLocalTime = cms.bool(True))
++>>>>>>> cms-sw/refs/pull/46010/head
  ticl_v5.toModify(hltTiclTrackstersCLUE3DHighL1Seeded.inferenceAlgo, type = cms.string('TracksterInferenceByDNN'))
diff --cc HLTrigger/Configuration/python/HLT_75e33/modules/hltTiclTrackstersCLUE3DHigh_cfi.py
index 88b4cd7b7ab,ee7df1e8e3e..00000000000
--- a/HLTrigger/Configuration/python/HLT_75e33/modules/hltTiclTrackstersCLUE3DHigh_cfi.py
+++ b/HLTrigger/Configuration/python/HLT_75e33/modules/hltTiclTrackstersCLUE3DHigh_cfi.py
@@@ -142,5 -142,5 +142,9 @@@ hltTiclTrackstersCLUE3DHigh = cms.EDPro
      )
      
  from Configuration.ProcessModifiers.ticl_v5_cff import ticl_v5
++<<<<<<< HEAD
 +ticl_v5.toModify(hltTiclTrackstersCLUE3DHigh.pluginPatternRecognitionByCLUE3D, computeLocalTime = cms.bool(True), doPidCut = cms.bool(False))
++=======
+ ticl_v5.toModify(hltTiclTrackstersCLUE3DHigh.pluginPatternRecognitionByCLUE3D, computeLocalTime = cms.bool(True))
++>>>>>>> cms-sw/refs/pull/46010/head
  ticl_v5.toModify(hltTiclTrackstersCLUE3DHigh.inferenceAlgo, type = cms.string('TracksterInferenceByDNN'))

```

you can try to solve the issue by doing this:

```
cmsrel CMSSW_14_2_X_2024-09-26-2300
cd CMSSW_14_2_X_2024-09-26-2300/src/
cmsenv
git-cms-fetch-pr 46010
git checkout pull/46010
git cms-addpkg HLTrigger/Configuration Configuration/PyReleaseValidation
git rebase -i CMSSW_14_2_X_2024-09-26-2300
# now fix manually the two conflicting files such that they are what they should be like
git commit
git rebase --continue
# continue fixing conflcts (rinse and repeat), when happy
git push my-cmssw +HEAD:dnnHLT
```

**Comment by RSalvatico** [comment]
It worked @mmusich, thank you!

**Comment by mmusich** [comment]
@cmsbuild, please test

File: HLTrigger/Configuration/python/HLT_75e33/modules/hltTiclTracksterLinksL1Seeded_cfi.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
+import FWCore.ParameterSet.Config as cms
+
+hltTiclTracksterLinksL1Seeded = cms.EDProducer("TracksterLinksProducer",
+    detector = cms.string('HGCAL'),
+    eid_input_name = cms.string('input'),
```
Comment (mmusich) [inline HLTrigger/Configuration/python/HLT_75e33/modules/hltTiclTracksterLinksL1Seeded_cfi.py:5]: something doesn't look right here, when comparing with the `fillDescriptions` of this producer:

https://github.com/cms-sw/cmssw/blob/eaefa4c3a28088f438f1d4b4d7a3839df859a143/RecoHGCal/TICL/plugins/TracksterLinksProducer.cc#L278-L303

I am getting runtime exceptions as:

```
----- Begin Fatal Exception 27-Sep-2024 18:45:08 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Validating configuration of module: class=TracksterLinksProducer label='hltTiclTracksterLinksSuperclusteringDNNL1Seeded'
Exception Message:
Illegal parameters found in configuration.  The parameters are named:
 'eid_input_name'
 'eid_min_cluster_energy'
 'eid_n_clusters'
 'eid_n_layers'
 'eid_output_name_energy'
 'eid_output_name_id'
 'tfDnnLabel'
You could be trying to use parameter names that are not
allowed for this plugin or they could be misspelled.
----- End Fatal Exception -------------------------------------------------
```

Comment (RSalvatico) [reply HLTrigger/Configuration/python/HLT_75e33/modules/hltTiclTracksterLinksL1Seeded_cfi.py:5]: Yes, the problem is that with the PR #45821 the TracksterLinksProducer has changed and so have the related configs, which I had looked at to create this one (and the unseeded version). Will change them accordingly.

**Comment by RSalvatico** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
ignore tests-rejected with external-failure
   * the issue with the phase-2 timing script is not due to this PR and is documented at https://github.com/cms-sw/cmssw/issues/46178

**Comment by mmusich** [comment]
For the record I tested privately and successfully this PR with the commands:
   * `hltPhase2UpgradeIntegrationTests --procModifiers ticl_v5 --restricPathsTo "HLT_Ele*" "HLT_Photon*"`
   * `hltPhase2UpgradeIntegrationTests --procModifiers ticl_v5,ticl_superclustering_mustache_ticl --restrictPathsTo "HLT_Ele*" "HLT_Photon*"`

I am still a bit puzzled by the results shown in the DQM bin-by-bin differences for wf 29691.204.
While there are now GSF electron tracks https://tinyurl.com/2c3x29ga, they don't seem to be associated with the PV https://tinyurl.com/25oupx4s. An indication of fakes?

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
Trying to test drive the recently introduced "maximum allocated memory" reporting, the maximum allocated memory in step2 of workflows 29691.203 and 29691.204 seems to increase by about 80-90 MB ([link](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-10-02-1100+eb053e/64855/maxmem-comparison/)). Is this along expectation?

**Comment by RSalvatico** [comment]
@mmusich I've run some tests and there's a couple of observations I can make:
- obvious comment: the reference offline track number is reduced by the requirement of compatibility with the PV of course (which is abs(dz) < 0.1, [here](https://github.com/cms-sw/cmssw/blob/b42e3a9d7bdebd44f56636c43adfc506845c38c1/DQM/TrackingMonitorSource/src/TrackToTrackComparisonHists.cc#L194-L196)), so in any case we start from a reduced sample of offline gsf tracks;
- I see the same behavior for the DNN superclustering (.203). In fact, working from a 14_1_0_pre7 release and looking at 10 events I get 6 offline gsf tracks compatible with the PV according the the above criterion. In the DNN case, I end up with 0 matched online tracks, in the mustache case with 2. What is interesting is that, even without the changes proposed in this PR, i.e. using a clean 14_1_0_pre7 release, the situation is similar: 0 matched gsf tracks in the .203 workflow and 4 matched gsf tracks in the .204 workflow.
So I believe whatever is the cause for the low number of matches, it goes deeper than this PR and did not start with it. I'll investigate further.

**Comment by RSalvatico** [comment]
If I am not wrong, the online-offline track association criterion is [deltaR < 0.002](https://github.com/cms-sw/cmssw/blob/b42e3a9d7bdebd44f56636c43adfc506845c38c1/DQM/TrackingMonitorSource/src/TrackToTrackComparisonHists.cc#L321) (applied [here](https://github.com/cms-sw/cmssw/blob/b42e3a9d7bdebd44f56636c43adfc506845c38c1/DQM/TrackingMonitorSource/src/TrackToTrackComparisonHists.cc#L211-L213)). Pretty stringent, but not much of a concern because the `ValidationWRTOffline/hltEgammaGsfTracks` is full, so evidently it's not too stringent. So let me take as an example an event with tracks that satisfy this matching:
- online eta = -2.45182, offline eta = -2.45182; online phi = -0.755262, offline phi = -0.755301. GenElectron eta = -2.45153, GenElectron phi = -0.755641. In this case the `abs(dz(PV)) < 0.1` criterion was not satisfied (because dZ(PV) = 0.540501).

There are several of such examples. I wouldn't say that these tracks are capturing fakes, but rather that the PV matching criterion is not often satisfied. Similar situation for the DNN (.203) workflow: out of 10 dielectron events, only two offline tracks passed the PV cutoff.

**Comment by RSalvatico** [comment]
> Trying to test drive the recently introduced "maximum allocated memory" reporting, the maximum allocated memory in step2 of workflows 29691.203 and 29691.204 seems to increase by about 80-90 MB ([link](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-10-02-1100+eb053e/64855/maxmem-comparison/)). Is this along expectation?

@makortel unfortunately I am not really sure how to quantify that from first principles... the only thing that comes to my mind is to compare it with the (possible) increase in maximum allocated memory in step3 following the PR #45333, which introduced similar changes offline. But I guess this is not feasible anymore, as that PR is merged.

**Comment by mmusich** [comment]
> the maximum allocated memory in step2 of workflows 29691.203 and 29691.204 seems to increase by about 80-90 MB 

in this PR `hltTiclTracksterLinksUnseeded` needs to load a couple of large model files (8-9MB) files from `ticlv5/onnx_models/linking` (on [cms-data](https://github.com/cms-data/RecoHGCal-TICL/tree/master/ticlv5/onnx_models/linking)) I guess this explains part of the memory allocation increase.

**Comment by mmusich** [comment]
test parameters:
   * enables hlt_p2_integration

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by RSalvatico** [comment]
As suggested by one of the authors of PR #45333 (@sameasy), I have disabled the supercluster energy regression in the `ticl_v5 mustache` case, since the regression version in release is not tuned for the mustache but only for the DNN supercluster. To be consistent, I have made this change online:
https://github.com/cms-sw/cmssw/blob/fe606281eafe348f85919b71be2f4561ce9b8916/HLTrigger/Configuration/python/HLT_75e33/sequences/HLTHgcalTiclPFClusteringForEgammaL1SeededSequence_cfi.py#L77

https://github.com/cms-sw/cmssw/blob/fe606281eafe348f85919b71be2f4561ce9b8916/HLTrigger/Configuration/python/HLT_75e33/sequences/HLTHgcalTiclPFClusteringForEgammaUnseededSequence_cfi.py#L98

and offline:
https://github.com/cms-sw/cmssw/blob/fe606281eafe348f85919b71be2f4561ce9b8916/RecoHGCal/TICL/python/superclustering_cff.py#L44

**Comment by RSalvatico** [comment]
@cmsbuild, please abort

**Comment by RSalvatico** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
test parameters:
   * workflow_opts= -w upgrade
   * workflows = 29691.203, 29691.204 
   * enable = hlt_p2_integration

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
+hlt 
   * for https://github.com/RSalvatico/cmssw/commit/013641436d50079a113d3021140407f1a21d5b26
   * PR according to description and follow-up review
   * phase-2 HLT integration tests are [passed](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-eb053e/42065/hlt-p2-integration.log)
   * private tests using the modifiers passed: https://github.com/cms-sw/cmssw/pull/46010#issuecomment-2385607578
   * no sizeable increase of timing: https://github.com/cms-sw/cmssw/pull/46010#issuecomment-2358356121
   * for a discussion of the changes in the DQM, see https://github.com/cms-sw/cmssw/pull/46010#issuecomment-2399043514 and following

**Comment by Moanwar** [comment]
+Upgrade

**Comment by AdrianoDee** [comment]
+pdmv

**Comment by RSalvatico** [comment]
> +4

Hi @jfernan2, one signature would be enough for me :)

**Comment by jfernan2** [comment]
+1

**Comment by mandrenguyen** [comment]
+1

**Pull request merged**